File: src/main/scala/net/machinemuse/powersuits/block/BlockTinkerTable.scala
package net.machinemuse.powersuits.block

import net.machinemuse.powersuits.common.{Config, ModularPowersuits}
import net.minecraft.block.{Block, SoundType}
import net.minecraft.block.material.Material
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.{EnumBlockRenderType, EnumFacing, EnumHand}
import net.minecraft.util.math.BlockPos
import net.minecraft.world.{IBlockAccess, World}
import net.minecraftforge.fml.common.registry.GameRegistry

/**
  * This is the tinkertable block. It doesn't do much except look pretty
  * (eventually) and provide a way for the player to access the TinkerTable GUI.
  *
  * @author MachineMuse
  *
  */

object BlockTinkerTable extends Block(Material.IRON) {
  this.setUnlocalizedName("tinkerTable")
  this.setHardness(1.5F)
  this.setResistance(1000.0F)
  this.setHarvestLevel("pickaxe", 2)
  this.setCreativeTab(Config.getCreativeTab)
  this.setSoundType(SoundType.METAL)
  this.setLightOpacity(0)
  this.setLightLevel(0.4f)
  this.setTickRandomly(false)

  var energyIcon: TextureAtlasSprite = null

  /**
    * Different render types do different things. Some will crash
    */
  override def getRenderType(state: IBlockState): EnumBlockRenderType = EnumBlockRenderType.MODEL

  /**
    * Called upon block activation (right click on the block.)
    */
  override def onBlockActivated(worldIn: World, pos: BlockPos, state: IBlockState, playerIn: EntityPlayer, hand: EnumHand, heldItem: ItemStack, side: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): Boolean = {
    if (playerIn.isSneaking) {
      false
    }
    if (worldIn.isRemote) {
      playerIn.openGui(ModularPowersuits.INSTANCE, 0, worldIn, pos.getX, pos.getY, pos.getZ)
    }
    true
  }

  /**
    * This method is called on a block after all other block gets already
    * created. You can use it to reference and configure something on the block
    * that needs the others ones.
    */
  protected def initializeBlock: Unit = {
  }

  /**
    * Returns Returns true if the given side of this block type should be
    * rendered (if it's solid or not), if the adjacent block is at the given
    * coordinates. Args: blockAccess, x, y, z, side
    */
  override def isBlockSolid(worldIn: IBlockAccess, pos: BlockPos, side: EnumFacing): Boolean = true

  /**
    * Is this block (a) opaque and (b) a full 1m cube? This determines whether
    * or not to render the shared face of two adjacent block and also
    * whether the player can attach torches, redstone wire, etc to this block.
    */
  override def isVisuallyOpaque: Boolean = false

  /**
    * Called throughout the code as a replacement for block instanceof
    * BlockContainer Moving this to the Block base class allows for mods that
    * wish
    * to extend vinella block, and also want to have a tile entity on that
    * block, may.
    *
    * Return true from this function to specify this block has a tile entity.
    *
    * @param state
    * IBlockState of the current block
    * @return True if block has a tile entity, false otherwise
    */
  override def hasTileEntity (state: IBlockState): Boolean = true

  /**
    * Called throughout the code as a replacement for
    * BlockContainer.getBlockEntity Return the same thing you would from that
    * function. This will
    * fall back to BlockContainer.getBlockEntity if this block is a
    * BlockContainer.
    *
    * @param state
    * The BlockState of the current block
    * @return A INSTANCE of a class extending TileEntity
    */
  override def createTileEntity (world: World, state: IBlockState): TileEntity = new TileEntityTinkerTable
}

File: src/main/java/net/machinemuse/powersuits/block/BlockTinkerTable.java
package net.machinemuse.powersuits.block;

import net.machinemuse.powersuits.tileentity.TileEntityTinkerTable;
import net.minecraft.block.Block;
import net.minecraft.block.BlockHorizontal;
import net.minecraft.block.SoundType;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.BlockFaceShape;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.BlockItemUseContext;
import net.minecraft.state.StateContainer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumBlockRenderType;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.World;
import net.minecraftforge.common.ToolType;

import javax.annotation.Nullable;

public class BlockTinkerTable extends BlockHorizontal {
    public BlockTinkerTable(String regName) {
        super(Block.Properties.create(Material.IRON)
                .hardnessAndResistance(1.5F, 1000.0F)
                .sound(SoundType.METAL)
                .variableOpacity()
                .lightValue(1));
        setRegistryName(regName);
        setDefaultState(this.stateContainer.getBaseState().with(HORIZONTAL_FACING, EnumFacing.NORTH));
    }

    @Override
    public int getHarvestLevel(IBlockState blockState) {
        return 2;
    }

    @Nullable
    @Override
    public ToolType getHarvestTool(IBlockState p_getHarvestTool_1_) {
        return ToolType.PICKAXE;
    }

    @Override
    protected void fillStateContainer(StateContainer.Builder<Block, IBlockState> builder) {
        builder.add(HORIZONTAL_FACING);
    }

    @Override
    public boolean hasTileEntity(IBlockState state) {
        return true;
    }

    @Nullable
    @Override
    public TileEntity createTileEntity(IBlockState state, IBlockReader world) {
        return new TileEntityTinkerTable();
    }

    /**
     * The type of render function called. MODEL for mixed tesr and static model, MODELBLOCK_ANIMATED for TESR-only,
     * LIQUID for vanilla liquids, INVISIBLE to skip all rendering
     * @deprecated call via {@link IBlockState#getRenderType()} whenever possible. Implementing/overriding is fine.
     */
    @Override
    public EnumBlockRenderType getRenderType(IBlockState state) {
        return EnumBlockRenderType.MODEL;
    }

    @SuppressWarnings( "deprecation" )
    @Override
    public BlockFaceShape getBlockFaceShape(IBlockReader worldIn, IBlockState state, BlockPos pos, EnumFacing face) {
        return BlockFaceShape.UNDEFINED;
    }

    @Override
    public boolean onBlockActivated(IBlockState state, World worldIn, BlockPos pos, EntityPlayer player, EnumHand hand, EnumFacing side, float hitX, float hitY, float hitZ) {
//        if (!worldIn.isRemote && !player.isSneaking())
//            NetworkHooks.openGui((EntityPlayerMP) player,new GuiTinkerTable(player, hitX, hitY, hitZ));
//
//            // see also minecraft.addScheduledTask(() -> minecraft.displayGuiScreen(gui));
//
//            ﻿
        return true;
    }

    @Nullable
    @Override
    public IBlockState getStateForPlacement(BlockItemUseContext context) {
        return this.getDefaultState().with(HORIZONTAL_FACING, context.getPlacementHorizontalFacing().getOpposite());
    }

    @Override
    public BlockRenderLayer getRenderLayer() {
        return BlockRenderLayer.CUTOUT;
    }

    @SuppressWarnings( "deprecation" )
    @Override
    public boolean isFullCube(IBlockState state) {
        return false;
    }
}